### ISSUE TITLE
[LLVM] nondeterministic stack overflow from lli with --jvm on bitcode compiled from C source (at -O2)

### ISSUE THREAD

**Issue by zhendongsu** [oracle/graal#4119]
GraalVM version or commit id if built from source: GraalVM CE 22.1.0-dev-20211209_2306 (the latest snapshot build from https://github.com/graalvm/graalvm-ce-dev-builds/releases)
CE or EE: CE
JDK version: openjdk 11.0.13
OS and OS Version: Ubuntu 18.04
Architecture: amd64
The output of `java -Xinternalversion`:
OpenJDK 64-Bit Server VM (11.0.13+8-jvmci-22.0-b02) for linux-amd64 JRE (11.0.13+8-jvmci-22.0-b02), built on Nov 22 2021 22:03:35 by "buildslave" with gcc 7.3.0

```
[547] % clangg22dev -v
GraalVM wrapper script for clang
GraalVM version: Development Build
running: /local/suz-local/software/graalvm-ce-java11-22.1.0-dev/lib/llvm/bin/clang -v
clang version 12.0.1 (GraalVM.org llvmorg-12.0.1-3-g6e0a5672bc-bgf11ed69a5a 6e0a5672bc058d882dce3d56f90b72b64a6870d7)
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /local/suz-local/software/graalvm-ce-java11-22.1.0-dev/lib/llvm/bin
Found candidate GCC installation: /usr/lib/gcc/i686-linux-gnu/8
Found candidate GCC installation: /usr/lib/gcc/x86_64-linux-gnu/6
Found candidate GCC installation: /usr/lib/gcc/x86_64-linux-gnu/6.5.0
Found candidate GCC installation: /usr/lib/gcc/x86_64-linux-gnu/7
Found candidate GCC installation: /usr/lib/gcc/x86_64-linux-gnu/7.5.0
Found candidate GCC installation: /usr/lib/gcc/x86_64-linux-gnu/8
Selected GCC installation: /usr/lib/gcc/x86_64-linux-gnu/7.5.0
Candidate multilib: .;@m64
Selected multilib: .;@m64
Found CUDA installation: /usr/local/cuda, version 11.0
[548] % 
[548] % lli22dev --version
LLVM 12.0.1 (GraalVM CE Native 22.1.0-dev)
[549] % 
[549] % clangg22dev -O2 small.c -o small
[550] % 
[550] % for i in $(seq 1 5) ; do echo $i; (ulimit -s unlimited; lli22dev --jvm small); done
1
2
Resource exhausted: Stack overflow
	at <llvm> main(Unknown)
3
4
Resource exhausted: Stack overflow
	at <llvm> main(Unknown)
5
[551] % 
[551] % cat small.c
char d[1][1], a[21], f;
int at[9][4][12][12][6], c, an, b;
void e() {
  for (char ag; b;)
#pragma clang loop unroll(enable)
    for (int aw = 0; aw < 13; aw += (char)56692274 - 49) {
      for (int ah = 0; ah < 10; ah += b)
        for (int ax = 0; ax < 18; ax += 4)
          for (char ai; ai < an; ai += 2)
            c = 0;
      for (char aj; aj < 3; aj += 2) {
        for (char ay; ay; ay++)
          d[ag][aw + aj] = 0;
#pragma clang loop vectorize_predicate(enable)
        for (int az = 0; az < 18; az += 2)
          for (int al = 0; al < 22; al += 2)
            a[aw + az * 19 + al] = 0;
      }
    }
}
int main() {
  for (unsigned af = 0; af < 9; ++af)
    for (unsigned ag = 0; ag < 4; ++ag)
      for (unsigned aw = 0; aw < 12; ++aw)
        for (unsigned ah = 0; ah < 12; ++ah)
          for (unsigned ax = 0; ag < 4; ++ag)
            for (; ax < 6; ++ax)
              at[af][ag][aw][ah][ax] = 1;
  for (unsigned af = 0; af < 9;)
    for (unsigned ag = 0; ag < 4;)
      for (unsigned aw = 0; aw < 3;)
        for (unsigned ah = 0; ah < 5;)
          for (unsigned ax; af < 20; ++af)
            for (; af < 7; ++af)
              for (; aw < 2;)
                for (ah = 0; ah < 9; ++ah)
                  for (; ag < 4; ++ag)
                    for (aw = 0; aw < 12; ++aw)
                      for (ah = ax = 0; ax < 1; ++ax)
                        f = 0;
  e();
  return 0;
}
```

Expected output: `lli22dev --jvm small` should execute normally without the error:
```
Resource exhausted: Stack overflow
	at <llvm> main(Unknown)
```

**Comment by rodrigar-mx** [comment]
Thanks for reporting this issue. We will take a look into it and get back to you.

**Comment by rodrigar-mx** [comment]
Hi @zhendongsu. I executed the code and I did not run into any issue. These were the steps I took:

1. $LLVM_TOOLCHAIN/clang  -O2 small.c -o small
2. for i in $(seq 1 5) ; do echo $i; (ulimit -s unlimited; $JAVA_HOME/bin/lli --jvm small); done
3. output:
   1
   2
   3
   4
   5

Is there any additional step?
I tested with GraalVM CE 22.1.0-dev-20211215_1015:
```
openjdk version "11.0.13" 2021-10-19
OpenJDK Runtime Environment GraalVM CE 22.1.0-dev (build 11.0.13+8-jvmci-22.0-b02)
OpenJDK 64-Bit Server VM GraalVM CE 22.1.0-dev (build 11.0.13+8-jvmci-22.0-b02, mixed mode, sharing)
```
